crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -4301,6 +4402,12 @@ impl<'db> TypeInferenceBuilder<'db> {
         let mut call_arguments = Self::parse_arguments(arguments);
         let callable_type = self.infer_expression(func);
 
+        if let Type::FunctionLiteral(function) = callable_type {
+            if function.definition(self.db()).scope(self.db()) == self.scope() {
+                self.called_functions.insert(function);
+            }
+        }
```

Comment:
This is the first part which adds the `FunctionType` that needs to be checked which is only added if it's defined and called in the same scope.